adr/0025-composition.md

```diff
@@ -0,0 +1,282 @@
+    # import:
+    #   path: common
+    # only:
+    #   flavor: registry1
+    #   cluster:
+    #     architecture: amd64
+    extends: mattermost
+    required: false
+
+    ## In this case (and as is the case with most `registry1` component flavors)
+    ## the Helm chart values were only used to override image tags. We can now
+    ## take advantage of a more robust and declarative `images` replacements API.
+    #
+    # charts:
+    #   - name: mattermost-enterprise-edition
+    #     valuesFiles:
+    #       - values/registry1-values.yaml
+    #       ## sample contents of `values/registry1-values.yaml`
+    #       # mattermostApp:
+    #       #   image:
+    #       #     repository: registry1.dso.mil/ironbank/opensource/mattermost/mattermost
+    #       #     tag: 9.9.0
+    #       # initContainerImage:
+    #       #   repository: registry1.dso.mil/ironbank/redhat/ubi/ubi9-minimal
+    #       #   tag: 9.4
+    # images:
+    #   - registry1.dso.mil/ironbank/redhat/ubi/ubi9-minimal:9.4
+    #   - registry1.dso.mil/ironbank/opensource/mattermost/mattermost:9.9.0
+    images:
+      - name: appropriate/curl
+        newName: registry1.dso.mil/ironbank/redhat/ubi/ubi9-minimal
+        newTag: 9.4
+      - name: mattermost/mattermost-enterprise-edition
+        newName: registry1.dso.mil/ironbank/opensource/mattermost/mattermost
+        newTag: 9.9.0
+
+  ## Finally, the most interesting example of *optionally* enabling injection of
+  ## Mattermost plugins. This component is now "stateless" because it does nothing
+  ## by default and it is "self-contained" because it both builds the required image
+  ## and includes the necessary Helm chart values overrides
+  ## (i.e. `mattermostApp.extraInitContainers`).
+  - name: mattermost-plugins
+    required: false
+
+    ## Note that the semantics of `requires` is slightly different from `extends`.
+    ## The idea is that `extends` signals intent to _replace_ the original component
+    ## whereas `requires` signals that the originaly component(s) are being overlaid
+    ## with additional configuration.
+    requires: [ mattermost ]
+    charts:
+      - name: mattermost-enterprise-edition
+        valuesFiles:
+          - values/extra-plugins-values.yaml
+          ## sample contents of `values/extra-plugins-values.yaml`:
+          #
+          # mattermostApp:
+          #   extraInitContainers:
+          #     - name: mattermost-extra-plugins
+          #       image: uds-package-mattermost/mattermost-extra-plugins:latest
+          #       imagePullPolicy: Always
+          #       volumeMounts:
+          #         - name: mattermost-plugins
+          #           mountPath: /mattermost/plugins/
+    images:
+      - name: uds-package-mattermost/mattermost-extra-plugins
+        newTag: latest
+    actions:
+      onCreate:
+        before:
+          - dir: plugins
+            cmd: |
+              docker buildx . -t uds-package-mattermost/mattermost-extra-plugins:latest
```

Comment:
The other way to solve this could be at the UDS CLI level to allow an additional image to be included alongside a Helm override (may still need an API change to the Zarf SDK to support but may be cleaner)

Reply:
(we could probably also get away from a docker build here ourselves and publish this image instead (and require others to do the same for theirs))

Reply:
For the purposes of this proposal, it doesn't matter whether we are doing a local docker build or not. This was just a self-contained way to demonstrate the need for an optional component to bundle both an additional image and related Helm chart value overrides.